metabase/metabase
# feat(sdk): font size scaling and adjustment for visualizations

Closes https://github.com/metabase/metabase/issues/42692

### Description

Before this, `theme.fontSize` used to only affect the SDK container's base font size. This PR makes the `theme.fontSize` settings affect the majority of viz typography, such as data tables, chart typography, axis ticks and series labels. The typography should scale relative to the base font size.

This also adds in optional settings to adjust viz font sizes: `table.cell.fontSize` and `cartesian` properties. This is used internally to apply `px` units to the Metabase app, while applying `em` relative values so they scale relative to the `theme.fontSize` value. I've exposed this as it's useful to the user despite not being part of the original goals. We convert the `em` values to `px` for echarts to use in offset and padding calculation.

### How to verify

- Go to the Metabase app and open questions/dashboards. The default font sizes should NOT change.
- Go to the `test-chart-font-size-scaling` branch in the demo app.
- Go to the Kitchen Sink page.
- Go to `AppProvider.tsx` and change `theme.fontSize` to `12px`, `14px`, `16px`, `18px` and `20px`. Notice the changes to the viz.
- Change `table.cell.fontSize` and `cartesian.label.fontSize`, which should affect data tables and cartesian charts.

### Limitations

These typographic elements are not yet affected by font size scaling. This will be addressed in a future PR.
- Pie chart's ring label and center label
- Funnel chart's bottom label
- Pivot table's cell

### Demo

Unset `theme.fontSize` or set to `14px`

<img src="https://github.com/metabase/metabase/assets/4714175/e83e2f64-fd3e-4d48-b020-3fc7273daa1b" width="300">

`theme.fontSize` set to `18px` (+4px)

<img src="https://github.com/metabase/metabase/assets/4714175/5072ea3a-6298-48e1-b326-7683dfccb41e" width="300">

### Checklist

- [x] Tests have been added/updated to cover changes in this PR - visual tests (line & bar chart)
